components/script/dom/htmlselectelement.rs

```diff
@@ -385,3 +381,67 @@ impl VirtualMethods for HTMLSelectElement {
 impl FormControl for HTMLSelectElement {}
 
 impl Validatable for HTMLSelectElement {}
+
+struct ListOfOptions {
+    curr: Option<Root<Node>>,
+    in_group: bool,
+}
+
+impl Iterator for ListOfOptions {
+    type Item = Root<HTMLOptionElement>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (ret, curr, in_group) = find_next_option(&self.curr, self.in_group);
+        self.curr = curr;
+        self.in_group = in_group;
+        ret
+    }
+}
+
+fn find_next_option(curr: &Option<Root<Node>>, in_group: bool) -> (Option<Root<HTMLOptionElement>>, Option<Root<Node>>, bool) {
```

Comment:
Can't you just keep a single HTMLOptionElement in the iterator and decide whether to go up or down the tree depending on whether its parent is an HTMLSelectElement or an HTMLOptGroupElement?